blog/modules/ROOT/pages/class-as-a-function.adoc

```diff
@@ -0,0 +1,267 @@
+= Class-as-a-Function Pattern
+
+:use-link-attrs:
+
+// tag::byline[]
+++++
+<div class="blog-byline">
+++++
+by link:https://github.com/HT154[Josh Basch] on August 6th, 2024
+++++
+</div>
+++++
+// end::byline[]
+
+// tag::excerpt[]
+In many languages, function and method parameters may be assigned default values allowing the argument to be omitted in calls.
+Many languages also offer named parameters, which aid in API self-documentation.
+Pkl xref:main:language-reference:index.adoc#methods[methods] do not provide either feature, but one way to achieve similar results is the "class-as-a-function" (CaaF) pattern.
+// end::excerpt[]
+
+== Usage
+
+In the class-as-a-function pattern, a class is created that accepts "input" properties and has one or more `fixed` "output" properties.
```

Comment:
When using CaaF, a thing that always pleases me is that you can make "multi-directional" functions. Consider, for example:
```
class Signal {
  frequency: Number = 1 / wavelength
  wavelength: Number = 1 / frequency
}
```
`Signal` is "only" a bi-directional function; it captures both `frequencyFromWavelength` and `wavelengthFromFrequency`.

Mostly, what I'm saying, is that I'm starting to see the "intentional" naming of properties like `result` ever more as a smell. In this case; starting by telling folks to think of certain properties as "input" and certain others as "output" is actually quite limiting.

I recommend naming properties for what they _are_; not for how they fit into the intended process. This helps push folks to be declarative.